File: A00/ruby/main.rb
def is_prime(n)
  if n < 2
    return false
  end

  for m in 2..(n - 1)
    if n % m == 0
      return false
    end
  end

  return true
end

def get_primes(lower_bound, upper_bound)
  list = []

  for p in lower_bound..upper_bound
    if is_prime(p)
      list.append(p)
    end
  end

  return list
end

def get_prime_partitions(n, bound = 2)
  partitions = []
  primes = get_primes(bound, n)

  if n < 2
    return []
  end

  for p in primes
    if n == p
      partitions += [[p]]
    end

    smaller_partitions = get_prime_partitions(n - p, p + 1)

    for partition in smaller_partitions
      partition.append(p)
    end

    partitions = partitions + smaller_partitions
  end

  return partitions
end

def print_prime_partitions(n)
  partitions = get_prime_partitions(n)

  if partitions.length == 0
    puts "There are no prime partitions of #{n}"
  else
    for partition in partitions
      puts "#{n} = #{partition.join(" + ")} (#{partition.length})"
    end

    puts "#{n} has #{partitions.length} prime partitions"
  end
end

n = -5
print_prime_partitions(n)

File: A00/haskell/main.hs
import Data.List

-- Checks if n has divisors in 2..k
hasDivisors :: Integer -> Integer -> Bool
hasDivisors n k
  | k > 1   = mod n k == 0 || hasDivisors n (k - 1)
  | k <= 1  = False

-- Checks if n is a prime number
isPrime :: Integer -> Bool
isPrime n
  | n > 1   = not (hasDivisors n (n - 1))
  | n < 0   = isPrime (-n)
  | n == 1  = False
  | n == 0  = False

-- Get all primes in between m and n
getPrimes :: Integer -> Integer -> [Integer]
getPrimes m n
  | n < m                 = []
  | n >= m  && isPrime m  = m : getPrimes (m + 1) n
  | otherwise             = getPrimes (m + 1) n

-- Append a number to every partition
appendToPartitions :: Integer -> [[Integer]] -> [[Integer]]
appendToPartitions p = map (p :)

{-
DISCLAIMER:
I originally wrote appendToPartitions using pattern matching, as follows

appendToPartitions p [] = []
appendToPartitions p (y : ys) = (p : y) : appendToPartitions p ys

and the linter suggested rewriting it using the map function and using the eta
reduction from p ys = map (p :) ys.
-}

-- Get prime partitions that contain a given value p
getPrimePartitionsAux :: Integer -> Integer -> Integer -> [[Integer]]
getPrimePartitionsAux n lb p
  | p == n = [[p]]
  | p /= n = appendToPartitions p (getPrimePartitions (n - p) (p + 1))

-- Get prime partitions
getPrimePartitions :: Integer -> Integer -> [[Integer]]
getPrimePartitions n lb
  | n < 2 || null primes =  []
  | n >= 2 && not (null primes) = getPrimePartitionsAux n lb (head primes) ++ getPrimePartitions n (head primes + 1)
    where
      primes = getPrimes lb n

printPartitions :: Integer -> [[Integer]] -> IO ()
printPartitions n [] = do return ()
printPartitions n (y : ys) = do
  putStrLn (show n ++ " = " ++ sum ++ " (" ++ show (length y) ++ ")")
  printPartitions n ys
  where
    sum = intercalate " + " (map show y)

main :: IO ()
main = do
  printPartitions n partitions
  putStrLn (show n ++ " has " ++ show (length partitions) ++ " prime partitions")
  where
    n = 42
    partitions = getPrimePartitions n 2